Add options to tracked functions for cycle recovery

XFFXFF opened the pull request:
closes #331 

This pr ports the old salsa tests for cycle in a single thread, except for [cycle_disappears_durability](https://github.com/salsa-rs/salsa/blob/03a27a7054ee234a566231243dd5025edfbcdde6/tests/cycles.rs#L326), since we don't have the api that permits setting durability.  

~I haven't ported parallel related tests, which would be some work, wondering if we can merge this in first~

XFFXFF commented on salsa-2022-tests/tests/cycles.rs:
```diff
@@ -0,0 +1,425 @@
+    let input = MyInput::new(&mut db);
+    let cycle = extract_cycle(|| memoized_a(&db, input));
+    let expected = expect![[r#"
+        [
```
We haven't implemented `debug` for [`DependencyKey`](https://github.com/salsa-rs/salsa/blob/03a27a7054ee234a566231243dd5025edfbcdde6/components/salsa-2022/src/key.rs#L37)

nikomatsakis commented:
> I haven't ported parallel related tests, which would be some work, wondering if we can merge this in first

yes, let's do that, but let's also create issues for the remaining to-do items...

XFFXFF commented:
@nikomatsakis  I ported all old salsa tests for cycle recovery, except for [cycle_disappears_durability](https://github.com/salsa-rs/salsa/blob/03a27a7054ee234a566231243dd5025edfbcdde6/tests/cycles.rs#L326), I added a [issue](#339)  for this.  
I also added #340 for more human readble output from  `DependencyIndex.debug`

nikomatsakis reviewed (approved):
This is awesome, thanks!

nikomatsakis commented:
@bors r+

nikomatsakis commented:
Egads.

bors r+
